### PR TITLE
[v10] inventory control stream

### PR DESCRIPTION
**Pull request by fspmarshall**
Backport of https://github.com/gravitational/teleport/pull/12715

### CODE REVIEW

**Review by rosstimothy — APPROVED** [review]

**Review by espadolini — APPROVED** [review]

File: lib/inventory/store.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	idx := entry.ct.Inc() % uint64(len(entry.handles))
+	handle = entry.handles[int(idx)]
+	return handle, true
+}
+
+func (s *shard) iter(fn func(UpstreamHandle)) {
+	s.rw.RLock()
+	defer s.rw.RUnlock()
+	for _, entry := range s.m {
+		idx := entry.ct.Inc() % uint64(len(entry.handles))
```
Comment (strideynet) [inline lib/inventory/store.go:141]: I believe this code is at risk of a divide by zero panic.

**Review by strideynet — CHANGES_REQUESTED** [review]

File: lib/inventory/store.go (hunk 1 above)
Comment (espadolini) [reply lib/inventory/store.go:141]: Yeah, we should either have length 0 as a special case or we should never have a zero-length `handles` by removing the whole entry when the last handle is removed.

Comment (fspmarshall) [reply lib/inventory/store.go:141]: Yeah, messed up my removal logic at some point.  Fixed panic and updated test to cover this path.

**Review by strideynet — APPROVED** [review]
Thanks for fixing this.

File: lib/inventory/store.go (hunk 1 above)
Comment (zmb3) [reply lib/inventory/store.go:141]: @fspmarshall please make sure this fix makes it in to master.

Great catch @strideynet!

**Pull request merged**